File: vi/codes/typescript/chapter_backtracking/preorder_traversal_iii_template.ts
/**
 * File: preorder_traversal_iii_template.ts
 * Created Time: 2023-05-09
 * Author: Justin (xiefahit@gmail.com)
 */

import { type TreeNode } from '../modules/TreeNode';
import { arrToTree } from '../modules/TreeNode';
import { printTree } from '../modules/PrintUtil';

/* Kiểm tra xem trạng thái hiện tại có phải là lời giải hay không */
function isSolution(state: TreeNode[]): boolean {
    return state && state[state.length - 1]?.val === 7;
}

/* Ghi lại lời giải */
function recordSolution(state: TreeNode[], res: TreeNode[][]): void {
    res.push([...state]);
}

/* Kiểm tra xem lựa chọn có hợp lệ dưới trạng thái hiện tại hay không */
function isValid(state: TreeNode[], choice: TreeNode): boolean {
    return choice !== null && choice.val !== 3;
}

/* Cập nhật trạng thái */
function makeChoice(state: TreeNode[], choice: TreeNode): void {
    state.push(choice);
}

/* Khôi phục trạng thái */
function undoChoice(state: TreeNode[]): void {
    state.pop();
}

/* Giải thuật quay lui: Ví dụ 3 */
function backtrack(
    state: TreeNode[],
    choices: TreeNode[],
    res: TreeNode[][]
): void {
    // Kiểm tra xem đây có phải là lời giải hay không
    if (isSolution(state)) {
        // Ghi lại lời giải
        recordSolution(state, res);
    }
    // Duyệt qua tất cả các lựa chọn
    for (const choice of choices) {
        // Cắt tỉa: kiểm tra xem lựa chọn có hợp lệ hay không
        if (isValid(state, choice)) {
            // Thử: thực hiện lựa chọn, cập nhật trạng thái
            makeChoice(state, choice);
            // Tiến hành vòng lựa chọn tiếp theo
            backtrack(state, [choice.left, choice.right], res);
            // Quay lui: hoàn tác lựa chọn, khôi phục về trạng thái trước đó
            undoChoice(state);
        }
    }
}

// Driver Code
const root = arrToTree([1, 7, 3, 4, 5, 6, 7]);
console.log('\nInitialize binary tree');
printTree(root);

// Giải thuật quay lui
const res: TreeNode[][] = [];
backtrack([], [root], res);

console.log('\nOutput all paths from root node to node 7, requiring paths do not include nodes with value 3');
res.forEach((path) => {
    console.log(path.map((node) => node.val));
});

export {};

File: vi/codes/ruby/chapter_backtracking/preorder_traversal_iii_template.rb
=begin
File: preorder_traversal_iii_template.rb
Created Time: 2024-05-22
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

require_relative '../utils/tree_node'
require_relative '../utils/print_util'

### Kiểm tra xem trạng thái hiện tại có phải lời giải hay không ###
def is_solution?(state)
  !state.empty? && state.last.val == 7
end

### Ghi lại lời giải ###
def record_solution(state, res)
  res << state.dup
end

### Kiểm tra xem lựa chọn có hợp lệ ở trạng thái hiện tại hay không ###
def is_valid?(state, choice)
  choice && choice.val != 3
end

### Cập nhật trạng thái ###
def make_choice(state, choice)
  state << choice
end

### Khôi phục trạng thái ###
def undo_choice(state, choice)
  state.pop
end

### Quay lui: ví dụ 3 ###
def backtrack(state, choices, res)
  # Kiểm tra xem đây có phải lời giải hay không
  record_solution(state, res) if is_solution?(state)

  # Duyệt qua tất cả các lựa chọn
  for choice in choices
    # Cắt tỉa: kiểm tra xem lựa chọn có hợp lệ hay không
    if is_valid?(state, choice)
      # Thử: thực hiện lựa chọn, cập nhật trạng thái
      make_choice(state, choice)
      # Tiến hành vòng lựa chọn tiếp theo
      backtrack(state, [choice.left, choice.right], res)
      # Quay lui: hoàn tác lựa chọn, khôi phục trạng thái trước đó
      undo_choice(state, choice)
    end
  end
end

### Mã chương trình chính ###
if __FILE__ == $0
  root = arr_to_tree([1, 7, 3, 4, 5, 6, 7])
  puts "\nInitialize binary tree"
  print_tree(root)

  # Thuật toán quay lui
  res = []
  backtrack([], [root], res)

  puts "\nOutput all paths from root node to node 7, requiring paths do not include nodes with value 3"
  for path in res
    p path.map { |node| node.val }
  end
end
